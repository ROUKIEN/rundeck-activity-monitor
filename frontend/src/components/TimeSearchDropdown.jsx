import React, { useState } from 'react'
import TimeSinceInput from './search/TimeSinceInput'
import TimeRangeInput from './search/TimeRangeInput'

const TIME_MODE_RANGE = 'range'
const TIME_MODE_SINCE = 'since'

const TimeSearchDropdown = ({ beginDate, endDate, onConfirm = () => {} }) => {
  const [currentBeginDate, setCurrentBeginDate] = useState(beginDate)
  const [currentEndDate, setCurrentEndDate] = useState(endDate)
  const [since, setSince] = useState(0)
  const [timeMode, setTimeMode] = useState(TIME_MODE_SINCE)

  const onValueChange = e => {
    setTimeMode(e)
  }

  const handleSearch = () => {
    if (timeMode == TIME_MODE_SINCE) {
      const start = new Date(Date.now() - since)
      onConfirm({
        start,
        end: new Date(),
      })
    } else {
      onConfirm({
        start: currentBeginDate,
        end: currentEndDate,
      })
    }
  }

  return (
  <div className="navbar-item has-dropdown is-hoverable">
    <a className="navbar-link">
      Search
    </a>
    <div className="navbar-dropdown is-right">
      <span className="navbar-item">
        <div className="control">
          <label className="radio">
            <input
              type="radio"
              name="mode"
              value={TIME_MODE_SINCE}
              defaultChecked={timeMode === TIME_MODE_SINCE}
              onClick={e => onValueChange(e.target.value)}
              data-testid={`radio-${TIME_MODE_SINCE}`}
            ></input>&nbsp;Since
          </label>
          <label className="radio">
            <input
              type="radio"
              name="mode"
              value={TIME_MODE_RANGE}
              defaultChecked={timeMode === TIME_MODE_RANGE}
              onClick={e => onValueChange(e.target.value)}
              data-testid={`radio-${TIME_MODE_RANGE}`}
            ></input>&nbsp;Range
          </label>
        </div>
      </span>
      { timeMode == TIME_MODE_RANGE && (
        <TimeRangeInput
          beginDate={currentBeginDate}
          endDate={currentEndDate}
          onBeginSelect={v => setCurrentBeginDate(v)}
          onEndSelect={v => setCurrentEndDate(v)}
        ></TimeRangeInput>)
      }
      { timeMode == TIME_MODE_SINCE && <TimeSinceInput onChange={ e => setSince(e)}></TimeSinceInput> }
      <span className="navbar-item">
        <div className="field is-grouped is-grouped-centered">
          <p className="control">
            <button className="button is-primary" onClick={() => handleSearch()}>Search</button>
          </p>
        </div>
      </span>
    </div>
  </div>
  )
}

export default TimeSearchDropdown
