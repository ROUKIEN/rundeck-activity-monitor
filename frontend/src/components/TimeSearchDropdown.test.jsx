import React from 'react'
import renderer from 'react-test-renderer'
import { expect, test } from 'vitest'
import TimeSearchDropdown from './TimeSearchDropdown'

test('emits a custom time range', () => {
  const fn = vi.fn(payload => {
    expect(payload).toHaveProperty('start', new Date('2022-01-01T00:00:00.000Z'))
    expect(payload).toHaveProperty('end', new Date('2022-01-02T00:00:00.000Z'))
  })

  const tc = renderer.create(
    <TimeSearchDropdown
      beginDate={new Date('2022-04-10T00:00:00')}
      endDate={new Date('2022-04-11T00:00:00')}
      onConfirm={fn}
    ></TimeSearchDropdown>
  )

  expect(tc.toJSON()).toMatchSnapshot('Defaults to since mode')

  const selectOption = tc.root.findAllByType('option')
  expect(selectOption).toHaveLength(4)

  const modeRadioInputs = tc.root.findAllByProps({
    type:'radio'
  })
  expect(modeRadioInputs).toHaveLength(2)
  modeRadioInputs[0].props.onClick({target: {value: 'range'}})

  const rangeInputs = tc.root.findAllByProps({
    type: 'datetime-local'
  })
  rangeInputs[0].props.onChange({
    target: {
      value: '2022-01-01T00:00',
      validity: {
        valid: true
      }
    }
  })

  rangeInputs[1].props.onChange({
    target: {
      value: '2022-01-02T00:00',
      validity: {
        valid: true
      }
    }
  })

  expect(tc.toJSON()).toMatchSnapshot('Switched to range mode with custom range')

  const searchButton = tc.root.findByProps({
    className: 'button is-primary'
  })

  searchButton.props.onClick()

  expect(fn).toHaveBeenCalledOnce()
})

test('emits a time range from the past 4 hours in since mode', () => {
  const fn = vi.fn(payload => {
    expect(payload).toHaveProperty('start')
    expect(payload).toHaveProperty('end')

    const { start, end } = payload
    expect(start.getHours()).toEqual(end.getHours() - 4)
  })

  const tc = renderer.create(
    <TimeSearchDropdown
      beginDate={new Date('2022-04-10T00:00:00')}
      endDate={new Date('2022-04-11T00:00:00')}
      onConfirm={fn}
    ></TimeSearchDropdown>
  )

  const select = tc.root.findByType('select')

  select.props.onChange({ target: { value: 4* 60 * 60 * 1000 }})

  const searchButton = tc.root.findByProps({
    className: 'button is-primary'
  })

  searchButton.props.onClick()

  expect(fn).toHaveBeenCalledOnce()
})
