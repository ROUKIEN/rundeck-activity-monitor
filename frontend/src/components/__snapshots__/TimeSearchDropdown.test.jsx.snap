// Vitest Snapshot v1

exports[`emits a custom time range > Defaults to since mode 1`] = `
<div
  className="navbar-item has-dropdown is-hoverable"
>
  <a
    className="navbar-link"
  >
    Search
  </a>
  <div
    className="navbar-dropdown is-right"
  >
    <span
      className="navbar-item"
    >
      <div
        className="control"
      >
        <label
          className="radio"
        >
          <input
            data-testid="radio-since"
            defaultChecked={true}
            name="mode"
            onClick={[Function]}
            type="radio"
            value="since"
          />
           Since
        </label>
        <label
          className="radio"
        >
          <input
            data-testid="radio-range"
            defaultChecked={false}
            name="mode"
            onClick={[Function]}
            type="radio"
            value="range"
          />
           Range
        </label>
      </div>
    </span>
    <span
      className="navbar-item"
    >
      <div
        className="select is-fullwidth"
      >
        <select
          onChange={[Function]}
        >
          <option
            value={3600000}
          >
            Past 1 hour
          </option>
          <option
            value={14400000}
          >
            Past 4 hours
          </option>
          <option
            value={43200000}
          >
            Past 12 hours
          </option>
          <option
            value={86400000}
          >
            Past 1 day
          </option>
        </select>
      </div>
    </span>
    <span
      className="navbar-item"
    >
      <div
        className="field is-grouped is-grouped-centered"
      >
        <p
          className="control"
        >
          <button
            className="button is-primary"
            onClick={[Function]}
          >
            Search
          </button>
        </p>
      </div>
    </span>
  </div>
</div>
`;

exports[`emits a custom time range > Switched to range mode with custom range 1`] = `
<div
  className="navbar-item has-dropdown is-hoverable"
>
  <a
    className="navbar-link"
  >
    Search
  </a>
  <div
    className="navbar-dropdown is-right"
  >
    <span
      className="navbar-item"
    >
      <div
        className="control"
      >
        <label
          className="radio"
        >
          <input
            data-testid="radio-since"
            defaultChecked={false}
            name="mode"
            onClick={[Function]}
            type="radio"
            value="since"
          />
           Since
        </label>
        <label
          className="radio"
        >
          <input
            data-testid="radio-range"
            defaultChecked={true}
            name="mode"
            onClick={[Function]}
            type="radio"
            value="range"
          />
           Range
        </label>
      </div>
    </span>
    <span
      className="navbar-item"
    >
      <div
        className="field"
      >
        <label
          className="label"
          htmlFor="dti-From"
        >
          From
        </label>
        <div
          className="control"
        >
          <input
            className="input is-fullwidth datetime-input"
            data-testid="datetimeinput-From"
            id="dti-From"
            onChange={[Function]}
            title="2022-01-01T00:00"
            type="datetime-local"
            value="2022-01-01T00:00"
          />
        </div>
      </div>
    </span>
    <span
      className="navbar-item"
    >
      <div
        className="field"
      >
        <label
          className="label"
          htmlFor="dti-To"
        >
          To
        </label>
        <div
          className="control"
        >
          <input
            className="input is-fullwidth datetime-input"
            data-testid="datetimeinput-To"
            id="dti-To"
            onChange={[Function]}
            title="2022-01-02T00:00"
            type="datetime-local"
            value="2022-01-02T00:00"
          />
        </div>
      </div>
    </span>
    <span
      className="navbar-item"
    >
      <div
        className="field is-grouped is-grouped-centered"
      >
        <p
          className="control"
        >
          <button
            className="button is-primary"
            onClick={[Function]}
          >
            Search
          </button>
        </p>
      </div>
    </span>
  </div>
</div>
`;
